WIP: Menu Components Adjustments | Responsive HFG Component Alignment | CSS Cleanup

abaicus opened the pull request:
### Summary
- Adds Responsive Radio Buttons customizer control.
- Makes HFG Primary and Secondary menus components `Item Spacing` and `Item Height` responsive.
- Makes all HFG components `Align` option responsive.
- Fixes issue with menu vertical overflow.
- Brutal CSS Cleanup. 
- Mega menu style is enqueued **only** when the mega menu actually exists.
- AMP dropdowns structure is updated.

<!-- Please describe the changes you made. -->

### Will affect visual aspect of the product
<!-- It includes visual changes? -->
YES - some things got changed, the goal being keeping the CSS file size to a minimum.

### Test instructions
<!-- Describe how this pull request can be tested. -->

#1820
---
Spacing of Primary and Secondary menus are now controlled by the responsive controls. Vertical spacing can be controlled using item height.

#1777, #1631, #1721
---
Items spacing and height are now responsive. For when items are in the sidebar, the spacing can be controlled by the item height, while outside the sidebar, the spacing still works properly.


#1719
---
Alignment settings are now responsive [the CSS needed for this is now three times what it was before - this includes special handling for the menus, as those need more specific styles].

#1718
---
Vertical overflow of the menus should work fine inside the sidebar now.

Testing:
---
- `Item Height` and `Item Spacing` should properly migrate from old versions. 
- `Alignment` should properly migrate from old versions.
- `Item Height` / `Item Spacing` on menu components (Primary|Secondary) should work as expected.
- `Alignment` on all components should work as expected.
- Menus in sidebar should **NOT** have a vertical scrollbar.
- Visual regression will be **very important** as I've refactored big chunks of CSS.
- Mega Menu CSS was split from main stylesheet. It should still work if you add a `neve-mega-menu` item in the primary menu.
- AMP integration should be in fully working order (the dropdowns markup was changed).

What Breaks in this PR:
---
When the menus go on two lines, the spacing is **NOT** properly handled anymore as that would bring a whole new dimension of exponential growth to our CSS. Menus going on multiple rows will have an offset if they are aligned right.

![image](https://user-images.githubusercontent.com/15010186/88791894-1d51dc80-d1a3-11ea-9af3-32af394076fa.png)
 

What happened to the CSS: 
---
![CSS Cleanup](https://media1.tenor.com/images/37c63438a0a570c95a80b8fef3206e8c/tenor.gif?itemid=8139307)

<!-- Issues that this pull request closes. -->
Closes #1820
Closes #1777
Closes #1631
Closes #1721
Closes #1719
Closes #1718
<!-- Should look like this: `Closes #1, #2, #3.` . -->

cc: @selul @rodica-andronache Please let me know (in this thread, not on individual issues) if you see anything weird, or anything acts up. Any feedback is appreciated. :pray:

rodica-andronache commented:
@abaicus here's what I saw while testing
- [ ] on a fresh install, I imported the Demo Agency. 
The `Primary menu appears to be right-aligned on mobile in the customizer control by default`, while the menu is actually aligned to the left side on the frontend.
This is why, if I go and update the alignment value for Desktop only, the mobile menu is also changing to the right alignment.
By default, there is a mismatch between the alignment option and the actual alignment, on mobile
![1](https://user-images.githubusercontent.com/3266155/89032607-b3216f00-d33d-11ea-9f33-16e9c1cee2d7.png)

That's why the mobile menu changes when I update the alignment for desktop
![2](https://user-images.githubusercontent.com/3266155/89032613-b583c900-d33d-11ea-83ef-e14f8fc65914.png)

- [ ] for `MegaMenu`, both on Desktop and Mobile, headings that are `Custom Links` ( and not Pages ) don't have the same styling and alignment as the other ones.
![3](https://user-images.githubusercontent.com/3266155/89032798-16130600-d33e-11ea-92a0-101ee950786b.png)
![4](https://user-images.githubusercontent.com/3266155/89032807-190df680-d33e-11ea-8651-069417416cca.png)

- [ ] in `AMP`, items of the Primary Menu that have a `dropdown` submenu don't get the same typography controls as the other items ( font family, font size )
![5](https://user-images.githubusercontent.com/3266155/89032916-55415700-d33e-11ea-8987-d5a314eb2a22.png)

- [ ] `Item Hover Color` doesn't apply to the items in `Primary Menu`. The color applies just on the underline on hover, not the text of the items. Saw this while migrating from an old menu, but it looks to not work anyway.
![6](https://user-images.githubusercontent.com/3266155/89033011-8a4da980-d33e-11ea-8077-37084c25608c.png)

- [ ] for the Primary Menu added on mobile, outside the sidebar, if it has dropdowns, you can open the dropdown by clicking on the arrow next to it, but you can't close it back unless you click on the arrow again and after that click outside the menu.
![7](https://user-images.githubusercontent.com/3266155/89033177-e9132300-d33e-11ea-8cae-0c9040392ee2.png)

selul commented:
@pirate-bot test:visual

selul commented:
@pirate-bot test:visual

abaicus commented:
@rodica-andronache I've fixed the things you mentioned. Let me know if you find anything else. :+1:

rodica-andronache commented:
@abaicus here are other things I've seen:
- [ ] the `Line height` Options for both Primary Menu and Secondary Menu don't look to apply, even though in the code I can see the custom values. Not sure if that's how it's supposed to be or not. I've tried both `em` and `px` and nothing changes no matter the value.
![1](https://user-images.githubusercontent.com/3266155/89402504-a7f48780-d71f-11ea-9da9-d87f8ef38f75.png)
![2](https://user-images.githubusercontent.com/3266155/89402526-ae82ff00-d71f-11ea-91fd-5fe173cf7e25.png)

- [ ] there are still some diferences ( alignment, padding) between headings that are Custom Links and heading that are Pages in MegaMenu ( only on desktop ). I'm testing with the menu from here https://qa-neve.themeisle.com/neve/ on the Web Agency demo
![3](https://user-images.githubusercontent.com/3266155/89402739-fd309900-d71f-11ea-8979-4b6c54401d8f.png)

abaicus commented:
@rodica-andronache the Height setting takes priority over the line-height one. It does apply to the dropdowns when the menu is not inside the sidebar, where the height of the items is not taken into consideration.

`height` is just a more important property. Layout properties have priority vs font properties. Not really sure how we could deal with this.

One idea would be to drop the line-height in the font style settings and base all the items (dropdowns & items) on the height. I think this was the same before so I find it a bit out-of-scope for this particular PR.

---

I've fixed the other stuff. :+1:

abaicus commented:
One more thing: I've also handled #1736 here. I've rethought the propositions I've made [here](https://github.com/Codeinwp/neve/issues/1736) as they would have probably introduced some accessibility issues for the older browser that don't work well with `:hover` or `:focus-within` states.

Here's what I thought would be the best solution to "_reconcile both the goat and the cabbage_":
- I left the functionality just as it was before, except that now the dropdowns should close when you click/tap outside them. 
- When dropdowns are outside the sidebar, the caret doesn't rotate anymore.  

---

I also handled a bug where dropdowns that were outside the viewport on the left side of the screen would not get repositioned properly :see_no_evil:. Also should work fine on RTL (which was not necessarily accounted for before).

Let me know if you see anything wrong there. :dancing_men: 

@rodica-andronache

irinelenache commented:
@abaicus When using mega menu, items that are too long get out of screen on mobile. 
https://prnt.sc/tyyxwb 
On neve 2.7.6 the item goes multi line if it's too long.

irinelenache commented:
- [ ] Mega menu headings inherit item color setted only if they have a link. (Posts and Hello world label are links)
https://prnt.sc/tzftzs
https://prnt.sc/tzfv5n

abaicus commented:
@irinelenache Thanks for looking into this! :+1: 

I've fixed the things you pointed out:
---

- Menu icons coming from OBFX should be aligned properly;
- Menu items titles should be wrapped properly when they exceed the sidebar width;
- Megamenu headings should be aligned the same as the other items at all times;
- Megamenu headings should use the same color as the other menu items;

Other stuff that I have changed as it seemed to make more sense this way: 
---

- The height option has been transformed into `min-height`. That means that you can override it with the `line-height` and the menu items won't overflow their container if they go on multiple lines.

- Changed how the mega-menu was aligned on the front-end. I've added a new class to the builder component container when there's a primary menu inside. This class is used to set the item's position as static, allowing for more leeway in how we are able to align. Now it's aligned relative to the header row that contains it.

rodica-andronache commented:
@abaicus tested again and the only remaining things I've seen are:
- [ ] imported the Web Agency starter site and the dropdowns have a black border. Not sure where is it coming from
![1](https://user-images.githubusercontent.com/3266155/90398767-9c4f7c00-e0a2-11ea-9938-ff5b1444ea55.png)
- [ ] for mega menu, there is still a slight difference between headings that are links and those that are not. The font-size is smaller for heading that are not links
![2](https://user-images.githubusercontent.com/3266155/90398961-dde02700-e0a2-11ea-8f53-c9e3bf24f29e.png)

abaicus commented:
@rodica-andronache the things you mentioned should be fixed now. 

I've dropped the border inside dropdowns altogether (confirmed with Marius that it's ok to do so). This removes a bit of CSS and inline style. 

Let me know if there's something else you've noticed that might not be ok. :+1:

abaicus commented:
I want to add that #1884 will get fixed by this PR too. :+1:

rodica-andronache commented:
@abaicus categories `descriptions` in the Primary menu, are not centered and don't use the `Items Color` or any other color control.
![a](https://user-images.githubusercontent.com/3266155/90508424-6a045400-e160-11ea-9ea6-668f38c4779f.png)

abaicus commented:
@rodica-andronache Found another bug caused by me here. Values from normal ranges weren't properly converted for responsive ranges. Sent in another commit. 

I've tested on my end and everything should be ok.

Can you please take a look too? The controls that were translated from normal to responsive are the Navigation Menus components (Primary, Secondary, Footer) `Item Min Height` and `Item Spacing`.

Thank you, and sorry for the mishap :disappointed:

rodica-andronache commented:
@abaicus found another issue. When the Mobile Menu Content layout is `Toggle Dropdown`, a white space appears below the menu, both on Desktop  and Mobile
![1](https://user-images.githubusercontent.com/3266155/90644482-714d5f80-e23d-11ea-969c-da650b8f68e9.png)
![2](https://user-images.githubusercontent.com/3266155/90644501-75797d00-e23d-11ea-97f9-59ed2e96ec7f.png)

rodica-andronache commented:
If you add the `Menu Icon` component on Desktop, the Sidebar that opens up has the Primary Menu aligned to the right side, even though the alignment in customizer is set to the left side.
![1](https://user-images.githubusercontent.com/3266155/90645966-48c66500-e23f-11ea-83cb-ac988a973f61.png)

This is the Sidebar on mobile. The Primary Menu is correctly aligned to the left side.
![2](https://user-images.githubusercontent.com/3266155/90645975-4d8b1900-e23f-11ea-9203-7fce3b2e0372.png)

abaicus commented:
@rodica-andronache 

> When the Mobile Menu Content layout is Toggle Dropdown, a white space appears below the menu, both on Desktop and Mobile

I don't think fixing this is a priority while refactoring it on another branch [here](https://github.com/Codeinwp/neve/tree/mobile-content-enhancements) 🙈 .

> If you add the Menu Icon component on Desktop, the Sidebar that opens up has the Primary Menu aligned to the right side, even though the alignment in customizer is set to the left side.

Well, the same rule applies for both of them - same component, same alignment. I am sure that the desktop alignment is set to right in the example above, while the mobile alignment is to the left. 🙈 

By default mobile alignment is to the left, while on desktop is to the right as it was previously. Mobile style will not apply to the sidebar as long as you're on a desktop-sized viewport. It's the same component. That derives from the fact that both mobile and desktop use the same component instances (it can be confusing at times).

rodica-andronache commented:
@abaicus ok, got it

rodica-andronache commented:
@abaicus the `dropdown arrows on Desktop` are bigger now than in the current version.
Also, the `icons` ( from OBFX ) are a bit higher compared to the text, than before. Not sure if that's correct

This PR
![a](https://user-images.githubusercontent.com/3266155/91153916-8442b280-e6c9-11ea-997d-b9dd9f48c460.png)


Current version
![b](https://user-images.githubusercontent.com/3266155/91153926-860c7600-e6c9-11ea-8ac9-fa1202d63cec.png)

abaicus commented:
@rodica-andronache this should be fine now.
